### PR TITLE
HDDS-2531. Sonar : remove duplicate string literals in BlockOutputStream

### PR DESCRIPTION
**Pull request by shwetayakkali**
## What changes were proposed in this pull request?
Refactored code to fix issues prompted by Sonar.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2531
## How was this patch tested?
Syntax changes and logging format changes.

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 1]
```diff
@@ -391,26 +391,20 @@ private void watchForCommit(boolean bufferFull) throws IOException {
...
           // for standalone protocol, logIndex will always be 0.
           commitWatcher
               .updateCommitInfoMap(asyncReply.getLogIndex(), byteBufferList);
         }
         return e;
       }, responseExecutor).exceptionally(e -> {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "putBlock failed for blockID " + blockID + " with exception " + e
+        LOG.debug("putBlock failed for blockID {} with exception {}", blockID,  e
```
Comment (dineshchitlangia) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: checkstyle issue reported at this line.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 2]
```diff
@@ -622,11 +613,9 @@ private void writeChunkToContainer(ChunkBuffer chunk) throws IOException {
...
           "Unexpected Storage Container Exception: " + e.toString(), e);
     }
-    if (LOG.isDebugEnabled()) {
-      LOG.debug(
-          "writing chunk " + chunkInfo.getChunkName() + " blockID " + blockID
-              + " length " + effectiveChunkSize);
-    }
+    LOG.debug(
+        "writing chunk {} blockID {} length {}",
+        chunkInfo.getChunkName(), blockID, effectiveChunkSize);
```
Comment (dineshchitlangia) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: NIT: rewrite it as
      `LOG.debug("writing chunk {} blockID {} length {}",`
     `              chunkInfo.getChunkName(), blockID, effectiveChunkSize);`

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 3]
```diff
@@ -536,9 +530,9 @@ private void setIoException(Exception e) {
           "Unexpected Storage Container Exception: " + e.toString(), e);
       ioException.compareAndSet(null, exception);
     } else {
-      LOG.debug("Previous request had already failed with " + ioe.toString()
+      LOG.debug("Previous request had already failed with {}"
           + " so subsequent request also encounters"
-          + " Storage Container Exception ", e);
+          + " Storage Container Exception ", ioe.toString(), e);
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: Avoid eagerly calling `toString()` for log parameters.

```suggestion
          + " Storage Container Exception ", ioe, e);
```

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 4]
```diff
@@ -609,11 +603,8 @@ private void writeChunkToContainer(ChunkBuffer chunk) throws IOException {
         }
         return e;
       }, responseExecutor).exceptionally(e -> {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "writing chunk failed " + chunkInfo.getChunkName() + " blockID "
-                  + blockID + " with exception " + e.getLocalizedMessage());
-        }
+        LOG.debug("writing chunk failed {} blockID {} with exception{}",
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: Nit: lost space in message.

```suggestion
        LOG.debug("writing chunk failed {} blockID {} with exception {}",
```

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: Just 1 character too long...
```suggestion
        LOG.debug("putBlock failed for blockID {} with exception {}", blockID, e
```

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 5]
```diff
@@ -391,26 +391,20 @@ private void watchForCommit(boolean bufferFull) throws IOException {
...
               .updateCommitInfoMap(asyncReply.getLogIndex(), byteBufferList);
         }
         return e;
       }, responseExecutor).exceptionally(e -> {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "putBlock failed for blockID " + blockID + " with exception " + e
-                  .getLocalizedMessage());
-        }
+        LOG.debug("putBlock failed for blockID {} with exception {}",
```
Comment (bharatviswa504) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: We don't need {} for exceptionMessage

Comment (elek) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: This is just the message of the exception not the stack trace itself (as before). I think it's reasonable to keep it as it's just a debug log which helps to localize the problem, the original exception is kept and part of the new exception which is propagated. So the full exception will be printed out somewhere from the callers.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java [hunk 6]
```diff
@@ -609,11 +603,8 @@ private void writeChunkToContainer(ChunkBuffer chunk) throws IOException {
         }
         return e;
       }, responseExecutor).exceptionally(e -> {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "writing chunk failed " + chunkInfo.getChunkName() + " blockID "
-                  + blockID + " with exception " + e.getLocalizedMessage());
-        }
+        LOG.debug("writing chunk failed {} blockID {} with exception {}",
```
Comment (xiaoyuyao) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: This is part of the I/O path where we have seen perf issue without the if (LOG.isDebugEnabled()), which slf4j claim that parameterize log will solve. So my suggestion is to keep the  if (LOG.isDebugEnabled()) and ignore the Sonar warning if any here.

Comment (swagle) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java]: Yes, we need to keep the explicit check for log level. The lazy evaluation only applies to the toString() call, the expression will be fully evaluated if any provided as a param, e.g.: e.getLocalizedMessage() above.

**Comment by arp7** [comment]
This jira seems to be inactive for a while. Does the patch need updating?

**Comment by swagle** [comment]
> This jira seems to be inactive for a while. Does the patch need updating?

One minor edit needed: add if (LOG.isDebugEnabled())

**Comment by elek** [comment]
/pending "One minor edit needed: add if (LOG.isDebugEnabled())"

**Pull request closed without merging**